cpp/src/binary/jit/core/cache.cpp

```diff
@@ -0,0 +1,212 @@
+    std::lock_guard<std::mutex> lock(_program_cache_mutex);
+
+    return getCached(prog_name, program_map, 
+        [&](){
+            CUDF_EXPECTS( not cuda_source.empty(),
+                "Program not found in cache, Needs source string.");
+            return jitify_v2::Program(cuda_source,
+                                        given_headers,
+                                        given_options,
+                                        file_callback);
+        }
+    );
+}
+
+named_prog<jitify_v2::KernelInstantiation> cudfJitCache::getKernelInstantiation(
+    std::string const& kern_name,
+    named_prog<jitify_v2::Program> const& named_program,
+    std::vector<std::string> const& arguments)
+{
+    // Lock for thread safety
+    std::lock_guard<std::mutex> lock(_kernel_cache_mutex);
+
+    std::string prog_name = std::get<0>(named_program);
+    jitify_v2::Program& program = *std::get<1>(named_program);
+
+    // Make instance name e.g. "prog_binop.kernel_v_v_int_int_long int_Add"
+    std::string kern_inst_name = prog_name + '.' + kern_name;
+    for ( auto&& arg : arguments ) kern_inst_name += '_' + arg;
+
+    return getCached(kern_inst_name, kernel_inst_map, 
+        [&](){return program.kernel(kern_name)
+                            .instantiate(arguments);
+        }
+    );
+}
+
+// Another overload for getKernelInstantiation which might be useful to get
+// kernel instantiations in one step
+// ------------------------------------------------------------------------
+/*
+jitify_v2::KernelInstantiation cudfJitCache::getKernelInstantiation(
+    std::string const& kern_name,
+    std::string const& prog_name,
+    std::string const& cuda_source = "",
+    std::vector<std::string> const& given_headers = {},
+    std::vector<std::string> const& given_options = {},
+    file_callback_type file_callback = nullptr)
+{
+    auto program = getProgram(prog_name,
+                              cuda_source,
+                              given_headers,
+                              given_options,
+                              file_callback);
+    return getKernelInstantiation(kern_name, program);
+}
+*/
+
+cudfJitCache::cacheFile::cacheFile(std::string file_name)
+ : _file_name{file_name}
+{ }
+
+cudfJitCache::cacheFile::~cacheFile() { }
+
+std::string cudfJitCache::cacheFile::read()
+{
+    // Open file (duh)
+    int fd = open ( _file_name.c_str(), O_RDWR );
+    if ( fd == -1 ) {
+        successful_read = false;
+        return std::string();
+    }
+
+    // Lock the file descriptor. we the only ones now
+    if ( lockf(fd, F_LOCK, 0) == -1 ) {
+        successful_read = false;
+        return std::string();
+    }
+
+    // Get file descriptor from file pointer
+    FILE *fp = fdopen( fd, "rb" );
+
+    // Get file length
+    fseek( fp , 0L , SEEK_END);
+    size_t file_size = ftell( fp );
+    rewind( fp );
+
+    // Allocate memory of file length size
+    char *buffer = (char*) malloc( file_size+1 );
```

Comment:
Doing this with C++ IO should be much nicer and avoid having to use an intermediate buffer by reading the file directly into a `std::string` or `std::stringstream`.

Reply:
Actually no. If you see the earlier commits in this PR, specifically https://github.com/rapidsai/cudf/pull/1245/commits/17a0ed9ab884124605ab1d7810c6212f2bcb06b2, you'll see that's exactly what I wanted to do. Unfortunately, file locks are not cross platform and the locks I used require file descriptors which are also a Unix only thing. Therefore the standards didn't expose file descriptors in fstreams.

https://stackoverflow.com/questions/11558447/retrieving-file-descriptor-from-a-stdfstream

And nor is it allowed to get the FILE pointer from fstreams.

https://stackoverflow.com/questions/109449/getting-a-file-from-a-stdfstream

Reply:
Ah... okay, so it ultimately stems from the fact that we need to use `lockf`, which only works on file descriptors... and there is no standard way to get the file descriptor from a `std::ofstream`.

Makes sense. Bummer though.

Reply:
Does C++17 improve this at all? e.g. with std::filesystem?

Reply:
`std::filesystem` doesn't add a lock because not all platforms have a lock system.